File: implementations/python-flask/vulnerable_app.py
"""
SQL Injection Vulnerable Implementation - Python Flask with SQLite
INTENTIONALLY INSECURE - For educational purposes only
"""

from flask import Flask, request, jsonify
import sqlite3
import os

app = Flask(__name__)

# Database connection - using SQLite
DB_PATH = os.path.join(os.path.dirname(__file__), 'data', 'sqlite.db')

def get_db_connection():
    conn = sqlite3.connect(DB_PATH)
    conn.row_factory = sqlite3.Row  # Access columns by name
    return conn


@app.route('/api/login', methods=['POST'])
def login():
    """
    VULNERABLE: Uses string formatting to build SQL query
    Susceptible to SQL injection attacks
    """
    data = request.get_json()
    username = data.get('username', '')
    password = data.get('password', '')
    
    # VULNERABILITY: Direct string concatenation
    query = f"SELECT UserId, Username, Email, Role FROM Users WHERE Username = '{username}' AND PasswordHash = '{password}'"
    
    try:
        conn = get_db_connection()
        cursor = conn.cursor()
        cursor.execute(query)  # Executing unsanitized query
        
        row = cursor.fetchone()
        conn.close()
        
        if row:
            return jsonify({
                'success': True,
                'user': {
                    'id': row[0],
                    'username': row[1],
                    'email': row[2],
                    'role': row[3]
                }
            })
        else:
            return jsonify({'success': False, 'message': 'Invalid credentials'}), 401
            
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500


@app.route('/api/posts/search', methods=['GET'])
def search_posts():
    """
    VULNERABLE: Uses string formatting for search queries
    Susceptible to UNION-based SQL injection
    """
    search_term = request.args.get('q', '')
    
    # VULNERABILITY: Direct string concatenation
    query = f"SELECT PostId, Title, Content FROM Posts WHERE Title LIKE '%{search_term}%' OR Content LIKE '%{search_term}%'"
    
    try:
        conn = get_db_connection()
        cursor = conn.cursor()
        cursor.execute(query)  # Executing unsanitized query
        
        posts = []
        for row in cursor.fetchall():
            posts.append({
                'id': row[0],
                'title': row[1],
                'content': row[2]
            })
        
        conn.close()
        return jsonify({'success': True, 'posts': posts})
        
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500


@app.route('/api/health', methods=['GET'])
def health():
    """Health check endpoint"""
    return jsonify({'status': 'running', 'version': 'vulnerable'})


if __name__ == '__main__':
    print("=" * 60)
    print("WARNING: This is an INTENTIONALLY VULNERABLE application!")
    print("For educational purposes only - DO NOT use in production")
    print("=" * 60)
    app.run(debug=True, port=5001)

File: implementations/nodejs-express/vulnerable_app.js
/**
 * SQL Injection Vulnerable Implementation - Node.js Express (sql.js)
 * INTENTIONALLY INSECURE - For educational purposes only
 */

const express = require('express');
const initSqlJs = require('sql.js');
const fs = require('fs');
const path = require('path');

const app = express();
app.use(express.json());

const DB_FILE = path.join(__dirname, 'data', 'sqlite.db');
let SQL; // sql.js module
let db;  // in-memory DB instance

function loadDatabase() {
  const filebuffer = fs.readFileSync(DB_FILE);
  const u8 = new Uint8Array(filebuffer);
  return new SQL.Database(u8);
}

const PORT = 3001;

(async () => {
  SQL = await initSqlJs();
  db = loadDatabase();

  app.listen(PORT, () => {
      console.log('='.repeat(60));
      console.log('WARNING: This is an INTENTIONALLY VULNERABLE application!');
      console.log('For educational purposes only - DO NOT use in production');
      console.log('='.repeat(60));
      console.log(`Server running on http://localhost:${PORT}`);
  });
})();

app.post('/api/login', async (req, res) => {
    const { username, password } = req.body;

    // VULNERABILITY: Direct string interpolation
    const query = `SELECT UserId, Username, Email, Role FROM Users WHERE Username = '${username}' AND PasswordHash = '${password}'`;

    try {
        const stmt = db.prepare(query);
        const hasRow = stmt.step();

        if (hasRow) {
            const row = stmt.getAsObject();
            res.json({
                success: true,
                user: {
                    id: row.UserId,
                    username: row.Username,
                    email: row.Email,
                    role: row.Role
                }
            });
        } else {
            res.status(401).json({ success: false, message: 'Invalid credentials' });
        }
        stmt.free();
    } catch (error) {
        res.status(500).json({ success: false, error: error.message });
    }
});

app.get('/api/posts/search', async (req, res) => {
    const searchTerm = req.query.q || '';
    const query = `SELECT PostId, Title, Content FROM Posts WHERE Title LIKE '%${searchTerm}%' OR Content LIKE '%${searchTerm}%'`;

    try {
        const stmt = db.prepare(query);
        const posts = [];
        while (stmt.step()) {
            const row = stmt.getAsObject();
            posts.push({ id: row.PostId, title: row.Title, content: row.Content });
        }
        stmt.free();
        res.json({ success: true, posts });
    } catch (error) {
        res.status(500).json({ success: false, error: error.message });
    }
});

app.get('/api/health', (req, res) => {
    res.json({ status: 'running', version: 'vulnerable' });
});
